File: app/src/main/java/afkt/project/ui/activity/GPUFilterACVActivity.java
package afkt.project.ui.activity;

import android.app.Activity;
import android.content.Intent;
import android.graphics.Bitmap;
import android.view.View;
import android.widget.AdapterView;

import com.luck.picture.lib.PictureSelector;
import com.luck.picture.lib.config.PictureMimeType;
import com.luck.picture.lib.entity.LocalMedia;

import java.util.ArrayList;
import java.util.List;

import afkt.project.R;
import afkt.project.base.app.BaseActivity;
import afkt.project.databinding.ActivityGpuFilterBinding;
import afkt.project.model.bean.ACVFileBean;
import afkt.project.ui.adapter.GPUFilterACVAdapter;
import afkt.project.util.GPUFilterUtils;
import dev.engine.log.DevLogEngine;
import dev.other.picture.PictureSelectorUtils;
import dev.utils.app.HandlerUtils;
import dev.utils.app.ResourceUtils;
import dev.utils.app.image.ImageUtils;
import jp.co.cyberagent.android.gpuimage.filter.GPUImageToneCurveFilter;

/**
 * detail: GPU ACV 文件滤镜效果
 * @author Ttt
 */
public class GPUFilterACVActivity
        extends BaseActivity<ActivityGpuFilterBinding> {

    // 适配器
    private        GPUFilterACVAdapter gpuFilterACVAdapter;
    // ACV 文件集合
    private final  List<ACVFileBean>   listACVFiles = new ArrayList<>();
    // 图片 Bitmap
    private        Bitmap              selectBitmap;
    // 滤镜线程
    private static Runnable            filterThread;

    @Override
    public int baseLayoutId() {
        return R.layout.activity_gpu_filter;
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        filterThread = null;
    }

    @Override
    public void initValue() {
        super.initValue();

        // 设置滤镜线程
        filterThread = new Runnable() {
            @Override
            public void run() {
                setFilter();
            }
        };

        // 初始化数据
        listACVFiles.add(new ACVFileBean("August", "filter/August.acv"));
        listACVFiles.add(new ACVFileBean("Darker", "filter/Darker.acv"));
        listACVFiles.add(new ACVFileBean("Dream", "filter/Dream.acv"));
        listACVFiles.add(new ACVFileBean("Fornature", "filter/Fornature.acv"));
        listACVFiles.add(new ACVFileBean("Greens", "filter/Greens.acv"));
        listACVFiles.add(new ACVFileBean("Miami", "filter/Miami.acv"));
        // 设置适配器
        binding.vidAgfGallery.setAdapter(gpuFilterACVAdapter = new GPUFilterACVAdapter(this, listACVFiles));
        binding.vidAgfGallery.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(
                    AdapterView<?> parent,
                    View view,
                    int position,
                    long id
            ) {
                gpuFilterACVAdapter.setSelectPosition(position);
                // 延迟一会进行滤镜
                HandlerUtils.removeRunnable(filterThread);
                HandlerUtils.postRunnable(filterThread, 500);
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
            }
        });
        // 默认选中第一个
        binding.vidAgfGallery.setSelection(0);
    }

    @Override
    public void initListener() {
        super.initListener();
        binding.vidAgfSelectBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                // 初始化图片配置
                PictureSelectorUtils.PicConfig picConfig = new PictureSelectorUtils.PicConfig()
                        .setCompress(false).setMaxSelectNum(1).setCrop(false).setMimeType(PictureMimeType.ofImage())
                        .setCamera(true).setGif(false);
                // 打开图片选择器
                PictureSelectorUtils.openGallery(PictureSelector.create(mActivity), picConfig);
            }
        });
    }

    // ===========
    // = 图片回传 =
    // ===========

    @Override
    protected void onActivityResult(
            int requestCode,
            int resultCode,
            Intent data
    ) {
        super.onActivityResult(requestCode, resultCode, data);

        // 判断是否属于图片选择
        if (resultCode == Activity.RESULT_OK && data != null) {
            LocalMedia localMedia = PictureSelectorUtils.getSingleMedia(data);
            // 获取图片地址
            String imgPath = PictureSelectorUtils.getLocalMediaPath(localMedia, true);
            // 获取图片 Bitmap
            selectBitmap = ImageUtils.decodeFile(imgPath);
            // 设置图片滤镜
            setFilter();
        }
    }

    // ===========
    // = 滤镜处理 =
    // ===========

    /**
     * 设置滤镜效果
     */
    private void setFilter() {
        try {
            if (selectBitmap == null) return;
            // 获取选中的滤镜
            int position = binding.vidAgfGallery.getSelectedItemPosition();
            // 获取滤镜文件实体类
            ACVFileBean acvFileBean = gpuFilterACVAdapter.getItem(position);
            // 设置滤镜效果
            GPUImageToneCurveFilter gpuFilter    = GPUFilterUtils.getGPUImageToneCurveFilter(ResourceUtils.open(acvFileBean.acvPath));
            Bitmap                  bitmapFilter = GPUFilterUtils.getFilterBitmap(selectBitmap, gpuFilter);
            binding.vidAgfIgview.setImageBitmap(bitmapFilter);
        } catch (Exception e) {
            DevLogEngine.getEngine().eTag(TAG, e, "setFilter");
        }
    }
}

File: app/src/main/java/afkt/project/ui/activity/GPUFilterACVActivity.kt
package afkt.project.ui.activity

import afkt.project.R
import afkt.project.base.app.BaseActivity
import afkt.project.base.config.RouterPath
import afkt.project.databinding.ActivityGpuFilterBinding
import afkt.project.model.bean.ACVFileBean
import afkt.project.ui.adapter.GPUFilterACVAdapter
import afkt.project.utils.GPUFilterUtils.getFilterBitmap
import afkt.project.utils.GPUFilterUtils.getGPUImageToneCurveFilter
import android.content.Intent
import android.graphics.Bitmap
import android.view.View
import android.widget.AdapterView
import com.alibaba.android.arouter.facade.annotation.Route
import dev.engine.DevEngine
import dev.engine.media.MediaConfig
import dev.utils.app.HandlerUtils
import dev.utils.app.ResourceUtils
import dev.utils.app.UriUtils
import dev.utils.app.image.ImageUtils

/**
 * detail: GPU ACV 文件滤镜效果
 * @author Ttt
 */
@Route(path = RouterPath.GPUFilterACVActivity_PATH)
class GPUFilterACVActivity : BaseActivity<ActivityGpuFilterBinding>() {

    // 适配器
    private lateinit var gpuFilterACVAdapter: GPUFilterACVAdapter

    // ACV 文件集合
    private val listACVFiles: MutableList<ACVFileBean> = ArrayList()

    // 图片 Bitmap
    private var selectBitmap: Bitmap? = null

    companion object {
        // 滤镜线程
        private var filterThread: Runnable? = null
    }

    override fun baseLayoutId(): Int = R.layout.activity_gpu_filter

    override fun onDestroy() {
        super.onDestroy()
        filterThread = null
    }

    override fun initValue() {
        super.initValue()

        // 设置滤镜线程
        filterThread = Runnable { setFilter() }

        // 初始化数据
        listACVFiles.add(ACVFileBean("August", "filter/August.acv"))
        listACVFiles.add(ACVFileBean("Darker", "filter/Darker.acv"))
        listACVFiles.add(ACVFileBean("Dream", "filter/Dream.acv"))
        listACVFiles.add(ACVFileBean("Fornature", "filter/Fornature.acv"))
        listACVFiles.add(ACVFileBean("Greens", "filter/Greens.acv"))
        listACVFiles.add(ACVFileBean("Miami", "filter/Miami.acv"))

        // 设置适配器
        binding.vidGallery.adapter = GPUFilterACVAdapter(this, listACVFiles).also {
            gpuFilterACVAdapter = it
        }
        binding.vidGallery.onItemSelectedListener =
            object : AdapterView.OnItemSelectedListener {
                override fun onItemSelected(
                    parent: AdapterView<*>?,
                    view: View,
                    position: Int,
                    id: Long
                ) {
                    gpuFilterACVAdapter.setSelectPosition(position)
                    // 延迟一会进行滤镜
                    HandlerUtils.removeRunnable(filterThread)
                    HandlerUtils.postRunnable(filterThread, 500)
                }

                override fun onNothingSelected(parent: AdapterView<*>?) {}
            }
        // 默认选中第一个
        binding.vidGallery.setSelection(0)
    }

    override fun initListener() {
        super.initListener()
        binding.vidSelectBtn.setOnClickListener {
            // 初始化图片配置
            val config = MediaConfig()
                .setCompress(false).setMaxSelectNum(1).setCrop(false)
                .setMimeType(MediaConfig.MimeType.ofImage())
                .setCamera(true).setGif(false)
            // 打开图片选择器
            DevEngine.getMedia()?.openGallery(mActivity, config)
        }
    }

    // ==========
    // = 图片回传 =
    // ==========

    override fun onActivityResult(
        requestCode: Int,
        resultCode: Int,
        data: Intent?
    ) {
        super.onActivityResult(requestCode, resultCode, data)
        // 判断是否属于图片选择
        if (resultCode == RESULT_OK && data != null) {
            // 获取图片地址
            val imgPath = DevEngine.getMedia()?.getSingleSelectorPath(data, true)
            // 获取图片 Bitmap
            selectBitmap = if (UriUtils.isUri(imgPath)) {
                ImageUtils.decodeStream(
                    ResourceUtils.openInputStream(
                        UriUtils.getUriForString(imgPath)
                    )
                )
            } else {
                ImageUtils.decodeFile(imgPath)
            }
            // 设置图片滤镜
            setFilter()
        }
    }

    // ==========
    // = 滤镜处理 =
    // ==========

    /**
     * 设置滤镜效果
     */
    private fun setFilter() {
        try {
            if (selectBitmap == null) return
            // 获取选中的滤镜
            val position = binding.vidGallery.selectedItemPosition
            // 获取滤镜文件实体类
            val acvFileBean = gpuFilterACVAdapter.getItem(position)
            // 设置滤镜效果
            val gpuFilter = getGPUImageToneCurveFilter(ResourceUtils.open(acvFileBean.acvPath))
            val bitmapFilter = getFilterBitmap(this, selectBitmap, gpuFilter)
            binding.vidIv.setImageBitmap(bitmapFilter)
        } catch (e: Exception) {
            DevEngine.getLog()?.eTag(TAG, e, "setFilter")
        }
    }
}